docs/GENERATORS.md

```diff
@@ -0,0 +1,314 @@
+# Test Generators
+
+Test generators allow tracks to generate tests automatically without having to write them ourselves. Each test generator reads from the exercise's `canonical data`, which defines the name of the test, its inputs, and outputs. You can read more about exercism's approach to test suites [here](https://github.com/exercism/docs/blob/master/language-tracks/exercises/anatomy/test-suites.md).
+
+Generating tests automatically removes any sort of user error when creating tests. We want the tests to be accurate with respect to its canonical data. Test generation also makes it much easier to keep tests up to date. As the canonical data changes, the tests will be automatically updated when the generator for that test is run.
+
+An example of a canonical data file can be found [here](https://github.com/exercism/problem-specifications/blob/master/exercises/bob/canonical-data.json)
+
+## Common Terms
+
+When looking through the canonical data and the generator code base, we use a lot of common terminology. This list hopefully clarifies what they represent.
+
+* Canonical Data - Represents the entire test suite.
+* Canonical Data Case - A representation of a single test case.
+* Description - The name of the test.
+* Property - The method to be called when running the test.
+* Input - The input for the test case.
+* Expected - The expected value when running the test case.
+
+## Adding a simple generator
+
+Adding a test generator is straightforward. Simply add a class to the `Generators.fs` file in the `generators` folder with the name of the exercise (in PascalCase), and extend the `GeneratorExercise` abstract class.
+
+An example of a simple generator would be the Bob exercise. The source code can be found below, but you can also view it in the repository [here](https://github.com/exercism/fsharp/blob/master/generators/Generators.fs#L174).
+
+```fsharp
+type Bob() =
+    inherit GeneratorExercise()
+```
+
+This is a fully working generator, no other code needs to be written. However, it's simplicity stems from the fact that the test suite and the program itself are relatively trivial.
+
+## Adding a complex generator
+
+For many generators, the default behavior of the `GeneratorExercise` class suffices. However, for more _complex_ generators, one can override one or more of its virtual methods:
+
+### Method: RenderInput
+
+Render the input of a test method.
+
+#### Example
+
+The `Gigasecond` generator renders the input value as a (parenthesized) `DateTime` value:
```

Comment:
You didn't link to the corresponding code here (and other places below) like you did above for the Bob generator. Was that intentional?

Reply:
Yes that was intentional, as in the F# version, all the generators are in the same file. This means that to refer to a specific generator, one would have to refer to a specific line. However, each change to that generators file would require us to also change all the line-based references. This seems like a too much work for the benefits.